### ISSUE TITLE
Global variables withing Python I

### ISSUE THREAD

**Issue by fwimp** [mhasoba/TheMulQuaBio#26]
It is probably prudent to remove the introduction to global variables in python and replace it with a simple demonstration of scope alongside maybe a few tests of people's scope intuition.

Globals are nearly universally a terrible idea in python, I feel that introducing them early sets them up as "a valid way to do things", leading to confusion on the need for return statements from functions.

**Comment by fwimp** [comment]
An example for a scope intuition question would be

Consider the following code:
``` python
i = 1
x = 0
for i in range(10):
    x += 1
print(x)
print(i)
```
What are the values of `x` and `i`?

Now consider this code:
``` python
i = 1
x = 0
def incrementer(r):
    x = 0
    for i in range(r):
        x += 1
    return x
x = incrementer(10)
print(x)
print(i)
```
What are the values of x and i in this case?

**Comment by davidorme** [comment]
That’s a bit sweeping. I’m not disagreeing exactly but global variables within packages aren’t uncommon - defining enums and the like.

That’s very different from the examples you have where you have the same variable names and unexpected results, so perhaps just a bit of nuance on defining global constants?

D




On 20 Oct 2020, at 10:32, Francis Windram <notifications@github.com<mailto:notifications@github.com>> wrote:

This email from notifications@github.com<mailto:notifications@github.com> originates from outside Imperial. Do not click on links and attachments unless you recognise the sender. If you trust the sender, add them to your safe senders list<https://spam.ic.ac.uk/SpamConsole/Senders.aspx> to disable email stamping for this address.



It is probably prudent to remove the introduction to global variables in python and replace it with a simple demonstration of scope alongside maybe a few tests of people's scope intuition.

Globals are nearly universally a terrible idea in python, I feel that introducing them early sets them up as "a valid way to do things", leading to confusion on the need for return statements from functions.

**Comment by fwimp** [comment]
I agree that globals within packages are a thing that happens (though again I'm rarely a particular fan of them, preferring instead to use a cfg file which I then import into other places in the package and then modify from there) but that is really very advanced python programming.

I don't believe we should remove mention of them entirely, but move it to later when they have developed good habits around programming more generally.

They are a useful concept, but in 4 consecutive years now I have seen students get the wrong end of the stick and write some almost unintelligible and undebuggable code because they didn't know any better about globals (which is perfectly understandable if they're one of the early things you learn.)

I am proposing more a demotion of their importance to a passing mention that you *can* do this and in certain circumstances it *could* be a good idea. They are useful when used properly, but the nuance of that is a lesson in and of itself :)

**Comment by fwimp** [comment]
Also globals are not thread-safe, which I think is the reason I have a particular distaste for them :P

**Comment by davidorme** [comment]
That’s not a problem if they are only used for module constants? I get the point though - there are a whole bunch of things they _shouldn’t_ be used for.

D


On 20 Oct 2020, at 11:11, Francis Windram <notifications@github.com<mailto:notifications@github.com>> wrote:

This email from notifications@github.com<mailto:notifications@github.com> originates from outside Imperial. Do not click on links and attachments unless you recognise the sender. If you trust the sender, add them to your safe senders list<https://spam.ic.ac.uk/SpamConsole/Senders.aspx> to disable email stamping for this address.



Also globals are not thread-safe, which I think is the reason I have a particular distaste for them :P

**Comment by fwimp** [comment]
Absolutely true, and they're unlikely to try multithreading much this early tbf.

**Comment by mbrazeau** [comment]
Sorry, not really my business here but just a thought: do you discuss encapsulation at all as good programming practice? Seems like this is a simple and useful way to address this.

OK, I'll bugger off now.
